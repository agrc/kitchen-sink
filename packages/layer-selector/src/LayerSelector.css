.layer-selector-item {
  margin: 0 8px;
}
.layer-selector-item--text {
  overflow: hidden;
  -o-text-overflow: ellipsis;
  text-overflow: ellipsis;
  font-size: 90%;
  max-width: 100px;
  white-space: nowrap;
}
.layer-selector--item {
  display: flex;
  flex-direction: row;
  align-items: center;
  user-select: none;
}
.layer-selector-item {
  margin: 0.5em;
}
.layer-selector-item:first-child {
  margin-top: 0;
}
.layer-selector-item:last-child {
  margin-bottom: 0;
}
.layer-selector-item input {
  margin: 0 0.5em 0 0;
}
.layer-selector {
  background: #fff;
  pointer-events: auto;
  max-width: 150px;
}
.layer-selector--width {
  width: 8em;
}
.layer-selector__toggle {
  width: 32px;
  height: 32px;
  padding: 5px;
  display: block;
}
.layer-selector--layers {
  min-height: 36px;
  min-width: 36px;
  padding: 10px 5px;
  white-space: nowrap;
}
.layer-selector--separator {
  display: block;
  height: 1px;
  border: 0;
  border-top: 1px solid #eee;
  padding: 0;
  margin-top: 0.5em;
  margin-bottom: 0.5em;
}
.layer-selector--hidden {
  display: none;
}
