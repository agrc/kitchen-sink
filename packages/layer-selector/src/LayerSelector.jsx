import { useState, useEffect, useRef } from 'react';
import classNames from 'clsx';
import PropTypes from 'prop-types';
import { createDefaultTileInfo } from './TileInfo';
import { setTileInfosForApplianceLayers } from './Discover';
import icon from './layers.svg';
import Basemap from '@arcgis/core/Basemap';
import LOD from '@arcgis/core/layers/support/LOD';
import TileInfo from '@arcgis/core/layers/support/TileInfo';
import WebTileLayer from '@arcgis/core/layers/WebTileLayer';
import FeatureLayer from '@arcgis/core/layers/FeatureLayer';

const commonFactories = {
  FeatureLayer,
  WebTileLayer,
};

const ExpandableContainer = (props) => {
  const [expanded, setExpanded] = useState(props.expanded);

  const imageClasses = classNames('layer-selector__toggle', {
    'layer-selector--hidden': expanded,
  });

  const fromClasses = classNames({ 'layer-selector--hidden': !expanded });

  return (
    <div
      className="layer-selector"
      // onFocus={() => setExpanded(true)}
      // onBlur={() => setExpanded(false)}
      onMouseOver={() => setExpanded(true)}
      onMouseOut={() => setExpanded(false)}
      area-haspopup="true"
    >
      <input type="image" className={imageClasses} src={icon} alt="layers" />
      <form className={fromClasses}>{props.children}</form>
    </div>
  );
};

ExpandableContainer.propTypes = {
  children: PropTypes.object,
  expanded: PropTypes.bool,
};

const LayerSelectorItem = (props) => {
  const inputOptions = {
    type: props.layerType === 'baselayer' ? 'radio' : 'checkbox',
    name: props.layerType,
    value: props.id,
  };

  return (
    <div className="layer-selector-item radio checkbox">
      <label className="layer-selector--item">
        <input
          {...inputOptions}
          checked={props.selected}
          onChange={(event) => props.onChange(event, props)}
        />
        <span className="layer-selector-item--text">{inputOptions.value}</span>
      </label>
    </div>
  );
};

const imageryAttributionJsonUrl =
  'https://mapserv.utah.gov/cdn/attribution/imagery.json';

/**
 * Takes layer tokens from `applianceLayers` keys and resolves them to `layerFactory` objects with
 * `esri/layer/WebTiledLayer` factories.
 * @private
 * @param {string} layerType - the type of layer `overlay` or `baselayer`.
 * @param {string[]|layerFactory[]} layerFactories - An array of layer tokens or layer factories.
 * @returns {layerFactory[]} an array of resolved layer Factory objects.
 */
const createLayerFactories = (
  layerType,
  layerFactories,
  WebTiledLayer,
  quadWord,
  applianceLayers
) => {
  const resolvedInfos = [];
  layerFactories.forEach((li) => {
    if (
      typeof li === 'string' ||
      li instanceof String ||
      li.token ||
      typeof li.token === 'string' ||
      li.token instanceof String
    ) {
      const id = li.token || li;

      if (!quadWord) {
        console.warn(
          'layer-selector::You chose to use a layer token `' +
            id +
            '` without setting ' +
            'your `quadWord` from Discover. The requests for tiles will fail to ' +
            ' authenticate. Pass `quadWord` into the constructor of this widget.'
        );

        return false;
      }

      var layer = applianceLayers[id];

      if (!layer) {
        console.warn(
          'layer-selector::The layer token `' +
            id +
            '` was not found. Please use one of ' +
            'the supported tokens (' +
            Object.keys(applianceLayers).join(', ') +
            ') or pass in the information on how to create your custom layer ' +
            '(`{Factory, url, id}`).'
        );

        return false;
      }

      const linked = [layer.linked, li.linked].reduce((acc, value, index) => {
        if (value) {
          acc = acc.concat(value);
        }

        if (index === 1 && acc.length === 0) {
          return null;
        }

        return acc;
      }, []);

      resolvedInfos.push({
        Factory: WebTiledLayer,
        urlTemplate: layer.urlPattern,
        linked,
        id,
        selected: !!li.selected,
        copyright: layer.copyright,
        layerType,
        // TODO: not supported in 4.x WebTileLayer copyright
        // hasAttributionData: layer.hasAttributionData,
        // attributionDataUrl: layer.attributionDataUrl
      });
    } else {
      if (!Object.prototype.hasOwnProperty.call(li, 'layerType')) {
        li.layerType = layerType;
      }

      if (!li.selected) {
        li.selected = false;
      }

      if (typeof li.Factory === 'string') {
        li.Factory = commonFactories[li.Factory];

        if (!li.Factory) {
          throw new Error(`Unknown layer factory: ${li.Factory}`);
        }
      }

      resolvedInfos.push(li);
    }
  });

  return resolvedInfos;
};

const ConditionalWrapper = ({ condition, wrapper, children }) =>
  condition ? wrapper(children) : children;
const LayerSelector = (props) => {
  const [layers, setLayers] = useState({
    baseLayers: [],
    overlays: [],
  });
  const [linkedLayers, setLinkedLayers] = useState([]);
  const [managedLayers, setManagedLayers] = useState({});
  const selectorNode = useRef();

  const [opacity, setOpacity] = useState(1);
  useEffect(() => {
    for (const layerId in managedLayers) {
      const managedLayer = managedLayers[layerId];
      if (managedLayer.layer) {
        managedLayer.layer.opacity = opacity;
      }
    }
  }, [managedLayers, opacity]);

  useEffect(() => {
    const managedLayersDraft = { ...managedLayers };
    const layerItems = layers.baseLayers.concat(layers.overlays);

    layerItems.forEach((layerItem) => {
      let layerList = null;
      switch (layerItem.layerType) {
        case 'baselayer':
          if (!props.view.map.basemap?.baseLayers) {
            props.view.map.basemap = {
              baseLayers: [],
              id: 'layer-selector',
              title: 'layer-selector',
            };
          }

          layerList = props.view.map.basemap.baseLayers;
          break;
        case 'overlay':
          layerList = props.view.map.layers;
          break;
        default:
          throw new Error(`unknown layerType: ${layerItem.layerType}`);
      }

      if (layerItem.selected === false) {
        var managedLayer = managedLayersDraft[layerItem.id] || {};
        if (!managedLayer.layer) {
          managedLayer.layer = layerList.getItemAt(
            layerList.indexOf(layerItem.layer)
          );
        }

        if (managedLayer.layer) {
          layerList.remove(managedLayer.layer);
        }

        return;
      }

      if (Object.keys(managedLayersDraft).indexOf(layerItem.id) < 0) {
        managedLayersDraft[layerItem.id] = {
          layerType: layerItem.layerType,
        };
      }

      if (!managedLayersDraft[layerItem.id].layer) {
        managedLayersDraft[layerItem.id].layer = new layerItem.Factory(
          layerItem
        );
      }

      if (layerItem.selected === true) {
        if (!layerList.includes(managedLayersDraft[layerItem.id].layer)) {
          layerList.add(managedLayersDraft[layerItem.id].layer);
        }
      } else {
        layerList.remove(managedLayersDraft[layerItem.id].layer);
      }

      // When you set the zoom on a map view without a cached layer, it has no effect on the scale of the map.
      // This is a hack to re-apply the zoom after adding the first cached layer. This works because the
      // map view is an observable
      managedLayersDraft[layerItem.id].layer.when('loaded', () => {
        const currentScale =
          managedLayersDraft[layerItem.id].layer.tileInfo.lods[props.view.zoom]
            .scale;
        if (props.view.zoom > -1 && props.view.scale !== currentScale) {
          // eslint-disable-next-line no-self-assign
          props.view.zoom = props.view.zoom;
        }
      });

      setManagedLayers(managedLayersDraft);
    });
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [layers]);

  useEffect(() => {
    if (!props.baseLayers || props.baseLayers.length < 1) {
      console.warn(
        'layer-selector::`baseLayers` is null or empty. Make sure you have spelled it correctly ' +
          'and are passing it into the constructor of this widget.'
      );

      return;
    }
  }, [props.baseLayers]);

  useEffect(() => {
    const applianceLayerTemplates = {
      Imagery: {
        urlPattern: `https://discover.agrc.utah.gov/login/path/${props.quadWord}/tiles/utah/{level}/{col}/{row}`,
        hasAttributionData: true,
        copyright: 'Hexagon',
        attributionDataUrl: imageryAttributionJsonUrl,
      },
      Topo: {
        urlPattern: `https://discover.agrc.utah.gov/login/path/${props.quadWord}/tiles/topo_basemap/{level}/{col}/{row}`,
        copyright: 'UGRC',
      },
      Terrain: {
        urlPattern: `https://discover.agrc.utah.gov/login/path/${props.quadWord}/tiles/terrain_basemap/{level}/{col}/{row}`,
        copyright: 'UGRC',
      },
      Lite: {
        urlPattern: `https://discover.agrc.utah.gov/login/path/${props.quadWord}/tiles/lite_basemap/{level}/{col}/{row}`,
        copyright: 'UGRC',
      },
      'Color IR': {
        urlPattern: `https://discover.agrc.utah.gov/login/path/${props.quadWord}/tiles/naip_2018_nrg/{level}/{col}/{row}`,
        copyright: 'UGRC',
      },
      Hybrid: {
        urlPattern: `https://discover.agrc.utah.gov/login/path/${props.quadWord}/tiles/utah/{level}/{col}/{row}`,
        linked: ['Overlay'],
        copyright: 'Hexagon, UGRC',
        hasAttributionData: true,
        attributionDataUrl: imageryAttributionJsonUrl,
      },
      Overlay: {
        urlPattern: `https://discover.agrc.utah.gov/login/path/${props.quadWord}/tiles/overlay_basemap/{level}/{col}/{row}`,
        // no attribution for overlay layers since it just duplicates the base map attribution
      },
      'Address Points': {
        urlPattern: `https://discover.agrc.utah.gov/login/path/${props.quadWord}/tiles/address_points_basemap/{level}/{col}/{row}`,
      },
    };

    props.view.map.basemap = new Basemap();

    const defaultTileInfo = createDefaultTileInfo(LOD);
    const applianceLayers = setTileInfosForApplianceLayers(
      applianceLayerTemplates,
      defaultTileInfo,
      TileInfo
    );

    const baseLayers =
      createLayerFactories(
        'baselayer',
        props.baseLayers,
        WebTileLayer,
        props.quadWord,
        applianceLayers
      ) || [];
    let overlays = props.overlays || [];

    let defaultSelection = null;
    let hasHybrid = false;
    let linkedLayersBuilder = [];

    baseLayers.forEach((layer) => {
      if (layer.selected === true) {
        defaultSelection = layer;
      }

      if ((layer.id || layer.token) === 'Hybrid') {
        hasHybrid = true;
      }

      if (layer.linked) {
        linkedLayersBuilder = linkedLayersBuilder.concat(layer.linked);
      }
    });
    setLinkedLayers(linkedLayersBuilder);

    // set default basemap to index 0 if not specified by the user
    if (!defaultSelection && baseLayers.length > 0) {
      baseLayers[0].selected = true;
      defaultSelection = baseLayers[0];
    }

    // insert overlay to first spot because hybrid
    if (
      hasHybrid &&
      overlays[0] !== 'Overlay' &&
      overlays[0]?.id !== 'Overlay'
    ) {
      overlays.splice(0, 0, 'Overlay');
    }

    overlays =
      createLayerFactories(
        'overlay',
        overlays,
        WebTileLayer,
        props.quadWord,
        applianceLayers
      ) || [];

    // set visibility of linked layers to match
    if (defaultSelection?.linked && defaultSelection.linked.length > 0) {
      overlays.forEach((layer) => {
        if (defaultSelection.linked.includes(layer.id)) {
          layer.selected = true;
        }
      });
    }

    setLayers({
      baseLayers,
      overlays,
    });

    props.view.ui.add(selectorNode.current, props.position);
  }, [
    props.baseLayers,
    props.overlays,
    props.position,
    props.quadWord,
    props.view.map,
    props.view.ui,
  ]);

  const onItemChanged = (event, props) => {
    console.log('LayerSelector.onItemChanged', props);
    const overlays = layers.overlays;
    const baseLayers = layers.baseLayers;

    if (props.layerType === 'baselayer') {
      baseLayers.forEach((item) => {
        item.selected = item.id === props.id ? event.target.checked : false;
      });

      const selectedItem = baseLayers.filter((layer) => layer.selected)[0];

      overlays.forEach((item) => {
        if (linkedLayers.includes(item.id)) {
          item.selected = false;
        }
      });

      if (selectedItem.linked && selectedItem.linked.length > 0) {
        overlays.forEach((item) => {
          if (selectedItem.linked.includes(item.id)) {
            item.selected = true;
          }

          return item;
        });
      }
    } else if (props.layerType === 'overlay') {
      overlays.forEach((item) => {
        if (item.id === props.id) {
          item.selected = event.target.checked;
        }
      });
    }

    setLayers({
      overlays,
      baseLayers,
    });
  };

  return (
    <div ref={selectorNode}>
      <ConditionalWrapper
        condition={props.makeExpandable}
        wrapper={(children) => (
          <ExpandableContainer>{children}</ExpandableContainer>
        )}
      >
        <div className="layer-selector--layers">
          {layers.baseLayers.map((item, index) => (
            <LayerSelectorItem
              id={item.name || item.id || 'unknown'}
              layerType="baselayer"
              selected={item.selected}
              onChange={onItemChanged}
              key={index}
            />
          ))}
          {layers.overlays.length ? (
            <hr className="layer-selector-separator" />
          ) : null}
          {layers.overlays.map((item) => (
            <LayerSelectorItem
              id={item.name || item.id || 'unknown'}
              layerType="overlay"
              selected={item.selected}
              onChange={onItemChanged}
              key={item.id || item}
            />
          ))}
          {props.showOpacitySlider ? (
            <>
              <hr className="layer-selector-separator" />
              <input
                type="range"
                min="0"
                max="100"
                step="1"
                value={opacity * 100}
                onChange={(event) =>
                  setOpacity(parseFloat(event.target.value) / 100.0)
                }
              />
            </>
          ) : null}
        </div>
      </ConditionalWrapper>
    </div>
  );
};

LayerSelector.propTypes = {
  view: PropTypes.object.isRequired,
  quadWord: PropTypes.string,
  baseLayers: PropTypes.arrayOf(
    PropTypes.oneOfType([
      PropTypes.oneOf([
        'Hybrid',
        'Lite',
        'Terrain',
        'Topo',
        'Color IR',
        'Address Points',
        'Overlay',
        'Imagery',
      ]),
      PropTypes.shape({
        Factory: PropTypes.oneOfType([PropTypes.func, PropTypes.string])
          .isRequired,
        urlTemplate: PropTypes.string,
        url: PropTypes.string,
        id: PropTypes.string.isRequired,
        tileInfo: PropTypes.object,
        linked: PropTypes.arrayOf(PropTypes.string),
      }),
      PropTypes.shape({
        token: PropTypes.oneOf([
          'Hybrid',
          'Lite',
          'Terrain',
          'Topo',
          'Color IR',
          'Address Points',
          'Overlay',
        ]).isRequired,
        selected: PropTypes.bool,
        linked: PropTypes.arrayOf(PropTypes.string),
      }),
    ])
  ).isRequired,
  overlays: PropTypes.arrayOf(
    PropTypes.oneOfType([
      PropTypes.oneOf(['Address Points', 'Overlay']),
      PropTypes.shape({
        Factory: PropTypes.oneOfType([PropTypes.func, PropTypes.string])
          .isRequired,
        urlTemplate: PropTypes.string,
        url: PropTypes.string,
        id: PropTypes.string.isRequired,
        tileInfo: PropTypes.object,
        linked: PropTypes.arrayOf(PropTypes.string),
      }),
    ])
  ),
  position: PropTypes.oneOf([
    'bottom-leading',
    'bottom-left',
    'bottom-right',
    'bottom-trailing',
    'top-leading',
    'top-left',
    'top-right',
    'top-trailing',
  ]),
  makeExpandable: PropTypes.bool,
  layerType: PropTypes.string,
  id: PropTypes.string,
  showOpacitySlider: PropTypes.bool,
};

LayerSelector.defaultProps = {
  makeExpandable: true,
  position: 'top-right',
  showOpacitySlider: false,
};

LayerSelectorItem.propTypes = {
  onChange: PropTypes.func.isRequired,
  selected: PropTypes.bool.isRequired,
  layerType: PropTypes.oneOf(['baselayer', 'overlay']).isRequired,
  id: PropTypes.string.isRequired,
};

export { ExpandableContainer, LayerSelectorItem };
export default LayerSelector;
