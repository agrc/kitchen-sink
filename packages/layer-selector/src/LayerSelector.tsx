import {
  Collection,
  Dialog,
  DialogTrigger,
  Popover as RACPopover,
  Header,
} from 'react-aria-components';
import { LayersIcon } from 'lucide-react';
import type { PopoverProps } from 'react-aria-components';
import {
  Button,
  Radio,
  RadioGroup,
  CheckboxGroup,
  Checkbox,
} from '@ugrc/utah-design-system/src';
import MapView from '@arcgis/core/views/MapView';

type LayerFactory = {
  Factory: new () => __esri.Layer;
  url: string;
  id: string;
  opacity: number;
};
type SelectorOptions = {
  view: MapView;
  quadWord: string;
  baseLayers: Array<
    string | { token: string; selected: boolean } | LayerFactory
  >;
  overlays?: Array<string | LayerFactory>;
  position: 'top-left' | 'top-right' | 'bottom-left' | 'bottom-right';
};

const Popover = (props: PopoverProps) => {
  return (
    <RACPopover
      {...props}
      className={({ isEntering, isExiting }) =>
        `group min-w-48 max-w-sm overflow-y-auto rounded-lg bg-white px-3 py-2 ring-1 ring-black/10 drop-shadow-lg dark:bg-zinc-800 dark:ring-white/10 ${
          isEntering
            ? 'animate-in fade-in placement-top:slide-in-from-bottom-1 placement-bottom:slide-in-from-top-1 duration-500 ease-out'
            : ''
        } ${
          isExiting
            ? 'animate-out fade-out placement-top:slide-out-to-bottom-1 placement-bottom:slide-out-to-top-1 duration-150 ease-in'
            : ''
        } `
      }
    />
  );
};

export function LayerSelector({
  options,
  ...props
}: {
  options: SelectorOptions;
}) {
  return (
    <DialogTrigger {...props}>
      <div className="inline-flex max-w-fit border border-black bg-white dark:border-zinc-500 dark:bg-zinc-900">
        <Button
          aria-label="Map layers"
          className="px-1.5 outline-none focus-visible:ring-2 focus-visible:ring-white/75"
          variant="icon"
        >
          <LayersIcon className="block size-8 p-1" />
        </Button>
      </div>
      <Popover>
        <Dialog className="outline-none">
          <Header className="font-bold dark:text-white">Base maps</Header>
          <RadioGroup className="flex-1">
            <Collection items={options.baseLayers}>
              {(item) => (
                <Radio className="pl-2" value={item.name}>
                  {item.name}
                </Radio>
              )}
            </Collection>
          </RadioGroup>
          <Header className="pt-3 font-bold dark:text-white">Overlays</Header>
          <CheckboxGroup className="flex-1">
            <Collection items={props.options.overlays}>
              {(item) => (
                <Checkbox className="pl-2" value="test">
                  {item.name}
                </Checkbox>
              )}
            </Collection>
          </CheckboxGroup>
        </Dialog>
      </Popover>
    </DialogTrigger>
  );
}
