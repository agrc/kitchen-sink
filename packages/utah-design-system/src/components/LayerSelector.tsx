import Basemap from '@arcgis/core/Basemap';
import TileInfo from '@arcgis/core/layers/support/TileInfo';
import MapView from '@arcgis/core/views/MapView';
import {
  Button,
  Checkbox,
  CheckboxGroup,
  Radio,
  RadioGroup,
} from '@ugrc/utah-design-system';
import { LayersIcon } from 'lucide-react';
import { useEffect, useRef, useState } from 'react';
import type { DialogTriggerProps, PopoverProps } from 'react-aria-components';
import {
  Dialog,
  DialogTrigger,
  Header,
  Popover as RACPopover,
} from 'react-aria-components';
import {
  layerTokens,
  type BaseLayerConfigOrToken,
  type LayerConfigOrToken,
} from './LayerSelector.types';
import { getLayerFromToken } from './LayerSelector.utilities';

type SelectorOptions = {
  view: MapView;
  /** discover quad word, required if using appliance tokens */
  quadWord?: string;
  /** layers added to map.basemap.baseLayers */
  baseLayers: BaseLayerConfigOrToken[];
  /** layers added to map.operationalLayers */
  operationalLayers?: LayerConfigOrToken[];
  /** layers added to map.basemap.referenceLayers */
  referenceLayers?: LayerConfigOrToken[];
  /** option passed to view.ui.add(), defaults to top-right */
  position?: 'top-left' | 'top-right' | 'bottom-left' | 'bottom-right';
};

const Popover = (props: PopoverProps) => {
  return (
    <RACPopover
      {...props}
      className={({
        isEntering,
        isExiting,
      }: {
        isEntering: boolean;
        isExiting: boolean;
      }) =>
        `group min-w-48 max-w-sm overflow-y-auto rounded-lg bg-white px-3 py-2 ring-1 ring-black/10 drop-shadow-lg dark:bg-zinc-800 dark:ring-white/10 ${
          isEntering
            ? 'duration-500 ease-out animate-in fade-in placement-top:slide-in-from-bottom-1 placement-bottom:slide-in-from-top-1'
            : ''
        } ${
          isExiting
            ? 'duration-150 ease-in animate-out fade-out placement-top:slide-out-to-bottom-1 placement-bottom:slide-out-to-top-1'
            : ''
        } `
      }
    />
  );
};

function getLabel(configOrToken: LayerConfigOrToken) {
  return typeof configOrToken === 'string'
    ? configOrToken
    : configOrToken.label;
}

export type LayerSelectorProps = Omit<DialogTriggerProps, 'children'> & {
  options: SelectorOptions;
};

export function LayerSelector({ options, ...props }: LayerSelectorProps) {
  const node = useRef<HTMLDivElement>(null);
  const managedLayers = useRef<Record<string, __esri.Layer>>({});

  // validate options
  useEffect(() => {
    if (options.baseLayers.length < 1) {
      throw new Error('`options.baseLayers` must have at least one layer');
    }

    if (!options.view.map) {
      throw new Error('`options.view` must have a map');
    }

    for (const configOrToken of [
      ...(options.baseLayers || []),
      ...(options.referenceLayers || []),
      ...(options.operationalLayers || []),
    ]) {
      if (typeof configOrToken === 'string') {
        if (!Object.values(layerTokens).includes(configOrToken)) {
          throw new Error(
            `layer-selector::The layer token '${configOrToken}' was not found. Please use one of the supported tokens (${Object.values(
              layerTokens,
            ).join(', ')}) or pass in a LayerConfig object.`,
          );
        }
        if (!options.quadWord) {
          throw new Error(
            `layer-selector::You chose to use a layer token '${configOrToken}' without setting your 'quadWord' from Discover. The requests for tiles will fail to authenticate. Pass 'quadWord' into the options parameter.`,
          );
        }
      }
    }
  }, [options]);

  const [selectedRadioBtnLabel, setSelectedRadioBtnLabel] = useState<string>(
    getLabel(options.baseLayers[0]!),
  );

  const referenceAndOperationalConfigsOrTokens = [
    ...(options?.referenceLayers || []),
    ...(options?.operationalLayers || []),
  ];
  const [selectedCheckboxLabels, setSelectedCheckboxLabels] = useState<
    string[]
  >(() =>
    referenceAndOperationalConfigsOrTokens
      .filter(
        (configOrToken) =>
          typeof configOrToken === 'object' && configOrToken?.defaultSelected,
      )
      .map(getLabel),
  );

  // set up map
  useEffect(() => {
    const basemapId = 'layer-selector';

    if (!node.current || options.view.map.basemap?.id === basemapId) return;

    options.view.map.basemap = new Basemap({
      id: basemapId,
    });

    options.view.ui.add(node.current, options.position || 'top-right');
  }, [options.position, options.view.map, options.view.ui]);

  // toggle layer visibility
  useEffect(() => {
    const map = options.view.map;
    const toggleLayer = async (
      configOrToken: LayerConfigOrToken,
      label: string,
      visible: boolean,
      container: __esri.Collection<__esri.Layer>,
    ) => {
      let layer = managedLayers.current[label];
      if (!layer && visible) {
        if (typeof configOrToken === 'string') {
          layer = getLayerFromToken(configOrToken, options.quadWord!);
        } else {
          layer = configOrToken.function();
        }
        managedLayers.current[label] = layer;
        container.add(layer);
      } else if (layer) {
        layer.visible = visible;
      }

      if (
        visible &&
        container === map.basemap!.baseLayers &&
        'tileInfo' in layer! &&
        layer.tileInfo instanceof TileInfo &&
        options.view.ready
      ) {
        const newMaxZoomLevel =
          layer.tileInfo.lods[layer.tileInfo.lods.length - 1]!.level;
        if (options.view.zoom > newMaxZoomLevel) {
          await options.view.goTo({ zoom: newMaxZoomLevel });
        }

        options.view.constraints.maxZoom = newMaxZoomLevel;
      }
    };

    for (const configOrToken of options.baseLayers) {
      const label = getLabel(configOrToken);
      toggleLayer(
        configOrToken,
        label,
        label === selectedRadioBtnLabel,
        map.basemap!.baseLayers,
      );
      if (label === 'Hybrid') {
        toggleLayer(
          'Overlay',
          'Overlay',
          label === selectedRadioBtnLabel,
          map.basemap!.referenceLayers,
        );
      }
    }

    for (const configOrToken of options.referenceLayers || []) {
      // todo: handle layer ordering (I *think* that esri might already make sure that polygons are under line are under points...)
      const label = getLabel(configOrToken);
      toggleLayer(
        configOrToken,
        label,
        selectedCheckboxLabels.includes(label),
        map.basemap!.referenceLayers,
      );
    }

    for (const configOrToken of options.operationalLayers || []) {
      // todo: handle layer ordering (I *think* that esri might already make sure that polygons are under line are under points...)
      const label = getLabel(configOrToken);
      toggleLayer(
        configOrToken,
        label,
        selectedCheckboxLabels.includes(label),
        map.layers,
      );
    }
  }, [
    options.baseLayers,
    options.operationalLayers,
    options.quadWord,
    options.referenceLayers,
    options.view,
    options.view.map,
    selectedCheckboxLabels,
    selectedRadioBtnLabel,
  ]);

  return (
    <div ref={node} className="esri-widget">
      <DialogTrigger {...props}>
        <Button
          aria-label="Map layers"
          className="esri-widget--button size-8 text-[#6e6e6e] outline-none focus-visible:ring-2 focus-visible:ring-white/75 dark:text-white"
          variant="icon"
        >
          <LayersIcon className="block size-8 p-1" />
        </Button>
        <Popover>
          <Dialog className="outline-none">
            <Header className="font-bold dark:text-white">Base maps</Header>
            <RadioGroup
              className="mb-2 flex-1"
              value={selectedRadioBtnLabel}
              onChange={setSelectedRadioBtnLabel}
            >
              {options.baseLayers.map((configOrToken) => {
                const value = getLabel(configOrToken);

                return (
                  <Radio className="pl-2" value={value} key={value}>
                    {value}
                  </Radio>
                );
              })}
            </RadioGroup>
            <Header className="font-bold dark:text-white">Overlays</Header>
            <CheckboxGroup
              className="mb-2 flex-1"
              value={selectedCheckboxLabels}
              onChange={setSelectedCheckboxLabels}
            >
              {referenceAndOperationalConfigsOrTokens.map((configOrToken) => {
                const label = getLabel(configOrToken);

                return (
                  <Checkbox className="pl-2" value={label} key={label}>
                    {label}
                  </Checkbox>
                );
              })}
            </CheckboxGroup>
          </Dialog>
        </Popover>
      </DialogTrigger>
    </div>
  );
}
