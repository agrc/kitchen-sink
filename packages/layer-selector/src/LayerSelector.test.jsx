import { render } from '@testing-library/react';
import { describe, expect, it } from 'vitest';
import LayerSelector from './';

describe('LayerSelector tests', () => {
  it('LayerSelector snapshot', () => {
    const props = {
      view: {
        map: {},
        when: () => {},
        ui: {
          add: () => {},
        },
      },
      baseLayers: ['Lite', 'Terrain', 'Topo', 'Color IR'],
      quadWord: 'my-fake-quad-word',
    };
    const { asFragment } = render(<LayerSelector {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
