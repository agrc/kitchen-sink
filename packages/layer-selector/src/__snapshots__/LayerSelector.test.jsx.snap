// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LayerSelector tests > LayerSelector snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      aria-haspopup="true"
      class="layer-selector"
    >
      <input
        alt="layers"
        class="layer-selector__toggle"
        src="/packages/layer-selector/src/layers.svg"
        type="image"
      />
      <form
        class="layer-selector--hidden"
      >
        <div
          class="layer-selector--layers"
        >
          <div
            class="layer-selector-item radio checkbox"
          >
            <label
              class="layer-selector--item"
            >
              <input
                checked=""
                name="baselayer"
                type="radio"
                value="Lite"
              />
              <span
                class="layer-selector-item--text"
              >
                Lite
              </span>
            </label>
          </div>
          <div
            class="layer-selector-item radio checkbox"
          >
            <label
              class="layer-selector--item"
            >
              <input
                name="baselayer"
                type="radio"
                value="Terrain"
              />
              <span
                class="layer-selector-item--text"
              >
                Terrain
              </span>
            </label>
          </div>
          <div
            class="layer-selector-item radio checkbox"
          >
            <label
              class="layer-selector--item"
            >
              <input
                name="baselayer"
                type="radio"
                value="Topo"
              />
              <span
                class="layer-selector-item--text"
              >
                Topo
              </span>
            </label>
          </div>
          <div
            class="layer-selector-item radio checkbox"
          >
            <label
              class="layer-selector--item"
            >
              <input
                name="baselayer"
                type="radio"
                value="Color IR"
              />
              <span
                class="layer-selector-item--text"
              >
                Color IR
              </span>
            </label>
          </div>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
